sympy/printing/smtlib.py

```diff
@@ -0,0 +1,515 @@
+            atan: 'arctan',
+            atan2: 'arctan2',
+            sinh: 'sinh',
+            cosh: 'cosh',
+            tanh: 'tanh',
+            Min: 'min',
+            Max: 'max',
+            Pow: 'pow',
+
+            And: 'and',
+            Or: 'or',
+            Xor: 'xor',
+            Not: 'not',
+            ITE: 'ite',
+            Implies: '=>',
+        }
+    }
+
+    symbol_table: dict
+
+    def __init__(self, settings: dict = None, symbol_table=None):
+        settings = settings or {}
+        self.symbol_table = symbol_table or {}
+        Printer.__init__(self, settings)
+        self._known_types = dict(self._settings['known_types'])
+        self._known_constants = dict(self._settings['known_constants'])
+        self._known_functions = dict(self._settings['known_functions'])
+
+        for _ in self._known_types.values(): assert self._is_legal_name(_)
+        for _ in self._known_constants.values(): assert self._is_legal_name(_)
+        # for _ in self._known_functions.values(): assert self._is_legal_name(_)  # +, *, <, >, etc.
+
+    def _is_legal_name(self, s: str):
+        if not s: return False
+        if s[0].isnumeric(): return False
+        return all(_.isalnum() or _ == '_' for _ in s)
+
+    def _s_expr(self, op: str, args: typing.Union[list, tuple]) -> str:
+        args_str = ' '.join(
+            a if isinstance(a, str)
+            else self._print(a)
+            for a in args
+        )
+        return f'({op} {args_str})'
+
+    def _print_Function(self, e):
+        if e in self._known_functions:
+            op = self._known_functions[e]
+        elif type(e) in self._known_functions:
+            op = self._known_functions[type(e)]
+        elif type(type(e)) == UndefinedFunction:
+            op = e.name
+        else:
+            op = self._known_functions[e]  # throw KeyError
+
+        return self._s_expr(op, e.args)
+
+    def _print_Relational(self, e: Relational):
+        return self._print_Function(e)
+
+    def _print_BooleanFunction(self, e: BooleanFunction):
+        return self._print_Function(e)
+
+    def _print_Expr(self, e: Expr):
+        return self._print_Function(e)
+
+    def _print_Unequality(self, e: Unequality):
+        if type(e) in self._known_functions:
+            return self._print_Relational(e)  # default
+        else:
+            eq_op = self._known_functions[Equality]
+            not_op = self._known_functions[Not]
+            return self._s_expr(not_op, [self._s_expr(eq_op, e.args)])
+
+    def _print_Piecewise(self, e: Piecewise):
+        def _print_Piecewise_recursive(args: typing.Union[list, tuple]):
+            e, c = args[0]
+            if len(args) == 1:
+                assert (c is True) or isinstance(c, BooleanTrue)
+                return self._print(e)
+            else:
+                ite = self._known_functions[ITE]
+                return self._s_expr(ite, [
+                    c, e, _print_Piecewise_recursive(args[1:])
```

Comment:
Maybe it's different in smtlib, but in SymPy, a Piecewise represents an expression (i.e., it evaluates to a number), but ITE represents a boolean (the `if`, `then`, and `else` all represent booleans). I guess smtlib handles these uniformly with its typing system?

Reply:
Yep!  I noticed that!  Sympy's `ITE` has signature `(if: bool) -> (then: bool) -> (else: bool) -> (result: bool)` whereas `ITE` in SMT has a signature more like `(if: bool) -> (then: T) -> (else: T) -> (result: T)`.